### PR TITLE
feature to check that FREE0 is a byte

### PR DESCRIPTION
**Pull request by zkronos73**
Feature to ASSERT that FREE0 is a byte value. This operation is different than $0 because it loads only the first chunk, FREE0, and never fails. BYTE is an ASSERT, which means that it loads only FREE0 (the rest of the chunks could have any value); the FREE0 must be a BYTE (0-255). It's useful when we are sure that a malicious prover will not provide a value different than a byte.

```
$BYTE{getByte()} => A
$BYTE{A & 0xFF} => A
$BYTE  => A      :MLOAD(padding[2])
```

### CODE REVIEW

**Review by krlosMata — APPROVED** [review]
:smile_cat:

**Review by hecmas — COMMENTED** [review]
Corrected

File: src/sm/sm_main/sm_main_exec.js [hunk 1]
```diff
@@ -2376,8 +2376,27 @@ module.exports = async function execute(pols, input, rom, config = {}, metadata
...
+            if (ctx.FREE0 < 0n || ctx.FREE0 > 255n) {
+                throw new Error(`FREE0 must be a byte, but has value ${ctx.FREE0} at ${sourceRef}`);
+            }
+            let value = ctx.FREE0;
+            for (let index = 0; index < 8; ++index) {
+                pols.hJmpnCondValueBit[index][i] = value & 0x01n;
+                value = value >> 1n;
+            }
+            pols.isNeg[i] = 0n;
+            pols.lJmpnCondValueBit[9][i] = 0n;
```
[the remark this answers is not in the document]
Comment (hecmas) [reply src/sm/sm_main/sm_main_exec.js:2389]: Solved #282

File: src/sm/sm_main/sm_main_exec.js [hunk 2]
```diff
@@ -2376,8 +2376,27 @@ module.exports = async function execute(pols, input, rom, config = {}, metadata
...
+            pols.hJmpnCondValueBit[i] = 0n;
```
[the remark this answers is not in the document]
Comment (hecmas) [reply src/sm/sm_main/sm_main_exec.js:2390]: Solved #282

File: src/sm/sm_main/sm_main_exec.js [hunk 3]
```diff
@@ -2408,10 +2427,12 @@ module.exports = async function execute(pols, input, rom, config = {}, metadata
             }
             pols.JMPN[i] = 1n;
         } else {
-            pols.isNeg[i] = 0n;
-            pols.lJmpnCondValue[i] = 0n;
-            for (let index = 0; index < 9; ++index) {
-                pols.hJmpnCondValueBit[index][i] = 0n;
+            if (!l.free0IsByte) {
+                pols.isNeg[i] = 0n;
```
[the remark this answers is not in the document]
Comment (hecmas) [reply src/sm/sm_main/sm_main_exec.js:2431]: It is correct.

File: src/sm/sm_main/sm_main_exec.js [hunk 4]
```diff
@@ -2408,10 +2427,12 @@ module.exports = async function execute(pols, input, rom, config = {}, metadata
...
+                pols.lJmpnCondValue[i] = 0n;
```
[the remark this answers is not in the document]
Comment (hecmas) [reply src/sm/sm_main/sm_main_exec.js:2432]: It is correct.

**Pull request merged**